File: src/main/java/pers/neige/neigeitems/action/node/impl/ChanceParser.java
package pers.neige.neigeitems.action.node.impl;

import lombok.NonNull;
import lombok.val;
import org.jetbrains.annotations.Nullable;
import pers.neige.neigeitems.action.ActionContext;
import pers.neige.neigeitems.action.node.NodeParser;
import pers.neige.neigeitems.config.ConfigReader;
import pers.neige.neigeitems.manager.BaseActionManager;
import pers.neige.neigeitems.utils.ListUtils;
import pers.neige.neigeitems.utils.NumberParser;
import pers.neige.neigeitems.utils.StringUtils;

import java.util.concurrent.ThreadLocalRandom;

public class ChanceParser extends NodeParser {
    public ChanceParser(@NonNull BaseActionManager manager) {
        super(manager);
    }

    @Override
    public @NonNull String getId() {
        return "chance";
    }

    @Override
    public @Nullable String parse(
        @NonNull ActionContext context,
        @NonNull ConfigReader params
    ) {
        return handle(
            getParsedValue(context, params, "success"),
            getParsedValue(context, params, "total"),
            getParsedValue(context, params, "repeat"),
            getParsedValue(context, params, "min"),
            getParsedValue(context, params, "max")
        );
    }

    @Override
    public @Nullable String parse(
        @NonNull ActionContext context,
        @NonNull String params
    ) {
        val paramsList = StringUtils.split(params, '_', '\\', 5);
        return handle(
            paramsList.get(0),
            ListUtils.getOrNull(paramsList, 1),
            ListUtils.getOrNull(paramsList, 2),
            ListUtils.getOrNull(paramsList, 3),
            ListUtils.getOrNull(paramsList, 4)
        );
    }

    private @Nullable String handle(
        @Nullable String success,
        @Nullable String total,
        @Nullable String repeat,
        @Nullable String min,
        @Nullable String max
    ) {
        if (success == null) {
            warning("未指定成功概率");
            return null;
        }

        val successRate = parse(success, 0D, NumberParser::parseDouble, "{0} 并非数字, 无法用作成功概率");
        if (successRate == null) return null;

        val totalRate = parse(total, 1D, NumberParser::parseDouble, "{0} 并非数字, 无法用作总概率");
        if (totalRate == null) return null;

        val repeatTimes = parse(repeat, 1, NumberParser::parseInteger, "{0} 并非整数, 无法用作重复次数");
        if (repeatTimes == null) return null;

        int result = 0;
        for (int i = 0; i < repeatTimes; i++) {
            if (successRate > ThreadLocalRandom.current().nextDouble(0D, totalRate)) {
                result++;
            }
        }

        if (min != null && !min.isEmpty()) {
            val minValue = parse(min, 0, NumberParser::parseInteger, "{0} 并非整数, 无法用作最小值");
            if (minValue == null) return null;
            result = Math.max(result, minValue);
        }

        if (max != null && !max.isEmpty()) {
            val maxValue = parse(max, 0, NumberParser::parseInteger, "{0} 并非整数, 无法用作最大值");
            if (maxValue == null) return null;
            result = Math.min(result, maxValue);
        }

        return Integer.toString(result);
    }
}

File: src/main/kotlin/pers/neige/neigeitems/section/impl/ChanceParser.kt
package pers.neige.neigeitems.section.impl

import org.bukkit.OfflinePlayer
import org.bukkit.configuration.ConfigurationSection
import pers.neige.neigeitems.section.SectionParser
import pers.neige.neigeitems.utils.ScriptUtils.toRoundingMode
import pers.neige.neigeitems.utils.SectionUtils.parseSection
import java.util.concurrent.ThreadLocalRandom

/**
 * 概率节点解析器
 */
object ChanceParser : SectionParser() {
    override val id: String = "chance"

    override fun onRequest(
        data: ConfigurationSection,
        cache: MutableMap<String, String>?,
        player: OfflinePlayer?,
        sections: ConfigurationSection?
    ): String? {
        return handler(
            cache,
            player,
            sections,
            true,
            data.getString("success"),
            data.getString("total"),
            data.getString("repeat"),
            data.getString("min"),
            data.getString("max")
        )
    }

    override fun onRequest(
        args: List<String>,
        cache: MutableMap<String, String>?,
        player: OfflinePlayer?,
        sections: ConfigurationSection?
    ): String {
        return handler(
            cache,
            player,
            sections,
            false,
            args.getOrNull(0),
            args.getOrNull(1),
            args.getOrNull(2),
            args.getOrNull(3),
            args.getOrNull(4)
        ) ?: "<$id::${args.joinToString("_")}>"
    }

    /**
     * @param cache 解析值缓存
     * @param player 待解析玩家
     * @param sections 节点池
     * @param parse 是否对参数进行节点解析
     * @param successString 成功概率文本
     * @param totalString 总概率文本
     * @param repeatString 重复次数文本
     * @param minString 最小值文本
     * @param maxString 最大值文本
     * @return 解析值
     */
    private fun handler(
        cache: MutableMap<String, String>?,
        player: OfflinePlayer?,
        sections: ConfigurationSection?,
        parse: Boolean,
        successString: String?,
        totalString: String?,
        repeatString: String?,
        minString: String?,
        maxString: String?
    ): String? {
        // 获取成功概率
        val successRate = successString?.parseSection(parse, cache, player, sections)?.toDoubleOrNull() ?: 0.0
        // 获取总概率
        val totalRate = totalString?.parseSection(parse, cache, player, sections)?.toDoubleOrNull() ?: 1.0
        // 获取重复次数
        val repeatTimes = repeatString?.parseSection(parse, cache, player, sections)?.toIntOrNull() ?: 1
        // 获取大小范围
        val min = minString?.parseSection(parse, cache, player, sections)?.toIntOrNull()
        val max = maxString?.parseSection(parse, cache, player, sections)?.toIntOrNull()
        // 成功次数
        var result = 0
        repeat(repeatTimes) {
            if (successRate > ThreadLocalRandom.current().nextDouble(0.0, totalRate)) {
                result += 1
            }
        }
        // 大小范围限制
        if (min != null) {
            result = min.coerceAtLeast(result)
        }
        if (max != null) {
            result = max.coerceAtMost(result)
        }
        return result.toString()
    }
}